src/snippet/mod.rs

```diff
@@ -0,0 +1,207 @@
+use tokenizer::{TokenStream, Tokenizer, Token};
+use std::collections::BTreeMap;
+use Term;
+use Document;
+use Index;
+use schema::FieldValue;
+use schema::Value;
+use tokenizer::BoxedTokenizer;
+
+#[derive(Debug)]
+pub struct HighlightSection {
+    start: usize,
+    stop: usize,
+}
+
+impl HighlightSection {
+    fn new(start: usize, stop: usize) -> HighlightSection {
+        HighlightSection {
+            start,
+            stop
+        }
+    }
+}
+
+#[derive(Debug)]
+pub struct FragmentCandidate {
+    score: f32,
+    start_offset: usize,
+    stop_offset: usize,
+    num_chars: usize,
+    highlighted: Vec<HighlightSection>,
+}
+
+impl FragmentCandidate {
+
+    fn new(start_offset: usize, end_offset: usize) -> FragmentCandidate {
+        FragmentCandidate{score: 0.0,
+                          start_offset: start_offset,
+                          stop_offset: end_offset,
+                          num_chars: 0,
+                          highlighted: vec![]}
+    }
+
+    /// Updates `score` and `highlighted` fields of the objects.
+    ///
+    ///
+    fn calculate_score(&mut self, token: &Token, terms: &BTreeMap<String, f32>) {
+        if let Some(score) = terms.get(&token.text.to_lowercase()) {
+            self.score += score;
+            self.highlighted.push(HighlightSection{start: token.offset_from,
+                                                   stop: token.offset_to});
+        }
+    }
+}
+
+#[derive(Debug)]
+pub struct Snippet {
+    fragments: String,
+    highlighted: Vec<HighlightSection>,
+}
+
+const HIGHLIGHTEN_PREFIX:&str = "<b>";
+const HIGHLIGHTEN_POSTFIX:&str = "</b>";
+
+impl Snippet {
+
+    /// Returns a hignlightned html from the `Snippet`.
+    pub fn to_html(&self) -> String {
+        let mut html = String::new();
+        let mut start_from: usize = 0;
+
+        for item in self.highlighted.iter() {
+            html.push_str(&self.fragments[start_from..item.start]);
```

Comment:
Please encode the html entities in the original text. You can add [htmlescape](https://crates.io/crates/htmlescape) as a dependency.